app/Metadata/GitHubFunctions.php

```diff
@@ -51,48 +41,69 @@ private function trim($commit_id): string
 	}
 
 	/**
-	 * look at .git/HEAD and return the current branch.
-	 * Return false if the file is not readable.
+	 * Looks at .git/HEAD and returns the current branch.
 	 *
-	 * @return false|string
+	 * @return string the current branch
+	 *
+	 * @throws VersionControlException
 	 */
-	public function get_current_branch(): string|false
+	private function determine_current_branch(): string
 	{
-		// @codeCoverageIgnoreStart
-		$head_file = base_path('.git/HEAD');
-		$branch_ = file_get_contents($head_file);
-		// separate out by the "/" in the string
-		$branch_ = explode('/', $branch_, 3);
+		try {
+			$head_file = base_path('.git/HEAD');
+			$branch = file_get_contents($head_file);
+			if ($branch === false) {
+				throw new \RuntimeException('`file_get_contents` returned `false`');
+			}
+			$branch = explode('/', $branch, 3);
+			if ($branch === false) {
+				throw new \RuntimeException('`explode` returned `false`');
+			}
 
-		return trim($branch_[2]);
-		// @codeCoverageIgnoreEnd
+			return trim($branch[2]);
```

Comment:
The branch is usually the last array key, so this should fix CI issues similar to https://github.com/LycheeOrg/Lychee/pull/1274.

```suggestion
			return trim(end($branch));
```

Reply:
The problem here is that the result in #1274 wouldn't be a branch name. Which may be fine, but is likely to have consequences.

Reply:
Actually, it looks like the errors here are new behaviour as a result of this PR. Previously the error would have been caught and they'd just be set to `false`.

Reply:
Isn't this only used to pull the latest commits? Then this is possible as well - you can use `git pull` with commit references. (I don't know where this method is used elsewhere.)

Reply:
@qwerty287 thanks for your suggestion, but I would rather like to keep that out of this PR. The PR is already ways to huge and this method was not introduced by this PR. So this is an unrelated fix.

If you feel like it, you are welcome to push another PR. But as @d7415 pointed out, your suggested change might have consequences elsewhere which I cannot assess right now. Personally, I consider this whole code ugly as hell. It is part of the unfortunate web installer/updater which is broken in so many ways that I plan to replace it entirely in the near future.

Reply:
> The PR is already ways to huge and this method was not introduced by this PR.

Agreed.

> So this is an unrelated fix.

[I'm less sure about that](https://github.com/LycheeOrg/Lychee/pull/1217#discussion_r857148691)

Reply:
Damn it. You caught me. Let's see what I can do about it.

Reply:
OK. I believe I have ultimately fixed all these problems. :crossed_fingers: The problem was not throwing exceptions or using an non-existing index. This was fine.

The problem was that the classes `GitHubFunctions` and `JsonRequestFunctions` already executed code in their constructors which was capable of throwing exceptions _and_ `GitHubFunctions` was registered as a service class in `AppServiceProvider`. Hence, Laravel already tried to construct these classes while building the routing pipeline even if these objects were never needed. The freshly constructed object `GitHubFunctions` tried to fetch GIT information and if it failed it threw an exception, but the `AppServiceProvider` was (and is) not prepared to handle those exceptions.

Now, the re-written classes only fetch the GIT information when the corresponding getter method is called for the first time. This may still fail with an exception, but this is OK, because we are prepared to handle these exceptions.

Reply:
That makes sense to me!